epochStart/bootstrap/fromLocalStorage.go

```diff
@@ -0,0 +1,198 @@
+	e.baseData.lastEpoch, e.baseData.shardId, e.baseData.lastRound, errNotCritical = storageFactory.FindLatestDataFromStorage(
+		e.generalConfig,
+		e.marshalizer,
+		e.workingDir,
+		e.genesisNodesConfig.ChainID,
+		e.defaultDBPath,
+		e.defaultEpochString,
+		e.defaultShardString,
+	)
+	if errNotCritical != nil {
+		log.Debug("no epoch db found in storage", "error", errNotCritical.Error())
+	} else {
+		e.baseData.storageExists = true
+		log.Debug("got last data from storage",
+			"epoch", e.baseData.lastEpoch,
+			"last round", e.baseData.lastRound,
+			"last shard ID", e.baseData.lastRound)
+	}
+}
+
+func (e *epochStartBootstrap) prepareEpochFromStorage() (Parameters, error) {
+	args := storageFactory.ArgsNewOpenStorageUnits{
+		GeneralConfig:      e.generalConfig,
+		Marshalizer:        e.marshalizer,
+		WorkingDir:         e.workingDir,
+		ChainID:            e.genesisNodesConfig.ChainID,
+		DefaultDBPath:      e.defaultDBPath,
+		DefaultEpochString: e.defaultEpochString,
+		DefaultShardString: e.defaultShardString,
+	}
+	openStorageHandler, err := storageFactory.NewStorageUnitOpenHandler(args)
+	if err != nil {
+		return Parameters{}, err
+	}
+
+	unitsToOpen := make([]string, 0)
+	unitsToOpen = append(unitsToOpen, e.generalConfig.BootstrapStorage.DB.FilePath)
+	unitsToOpen = append(unitsToOpen, e.generalConfig.MetaBlockStorage.DB.FilePath)
+
+	storageUnits, err := openStorageHandler.OpenStorageUnits(unitsToOpen)
+	defer func() {
+		for _, storer := range storageUnits {
+			errClose := storer.Close()
+			log.LogIfError(errClose)
+		}
+	}()
+
+	if err != nil || len(storageUnits) != len(unitsToOpen) {
+		return Parameters{}, err
+	}
+
+	_, e.nodesConfig, err = e.getLastBootstrapData(storageUnits[0])
+	if err != nil {
+		return Parameters{}, err
+	}
+
+	pubKey, err := e.publicKey.ToByteArray()
+	if err != nil {
+		return Parameters{}, err
+	}
+
+	if !e.checkIfShuffledOut(pubKey, e.nodesConfig) {
+		parameters := Parameters{
+			Epoch:       e.baseData.lastEpoch,
+			SelfShardId: e.baseData.shardId,
+			NumOfShards: e.baseData.numberOfShards,
+		}
+		return parameters, nil
+	}
+
+	e.epochStartMeta, err = e.getEpochStartMetaFromStorage(storageUnits[1])
+	if err != nil {
+		return Parameters{}, err
+	}
+
+	err = e.createSyncers()
+	if err != nil {
+		return Parameters{}, err
+	}
+
+	e.syncedHeaders, err = e.syncHeadersFrom(e.epochStartMeta)
+	if err != nil {
+		return Parameters{}, err
+	}
+
+	prevEpochStartMetaHash := e.epochStartMeta.EpochStart.Economics.PrevEpochStartHash
+	prevEpochStartMeta, ok := e.syncedHeaders[string(prevEpochStartMetaHash)].(*block.MetaBlock)
+	if !ok {
+		return Parameters{}, epochStart.ErrWrongTypeAssertion
+	}
+	e.prevEpochStartMeta = prevEpochStartMeta
+
+	e.shardCoordinator, err = sharding.NewMultiShardCoordinator(e.baseData.numberOfShards, e.baseData.shardId)
+	if err != nil {
+		return Parameters{}, err
+	}
+
+	if e.shardCoordinator.SelfId() == core.MetachainShardId {
+		err = e.requestAndProcessForShard()
```

Comment:
If meta, `requestAndProcessForShard` / from shard? Just double check if this is correct.

Reply:
fixed.